src/mame/jaleco/jaleco_vj_qtaro.cpp

```diff
@@ -0,0 +1,653 @@
+            const uint32_t flags = m_dma_space->read_dword(m_dma_descriptor_addr[device_id] + 12); // Bit 24 is set to denote the last entry at the same time as bit 0 of the next descriptor addr is set
+
+            LOGMASKED(LOG_DMA, "DMA %d: %08x -> %08x %08x (%d %d)\n", device_id, m_dma_descriptor_addr[device_id], nextDescriptorPhysAddr, m_dma_descriptor_length[device_id], BIT(nextDescriptorPhysAddr, 0), BIT(flags, 24));
+
+            m_dma_descriptor_addr[device_id] = nextDescriptorPhysAddr;
+            m_dma_descriptor_length[device_id] = 0;
+            m_dma_running[device_id] = BIT(m_dma_descriptor_addr[device_id], 0) == 0 && BIT(flags, 24) == 0;
+        }
+    }
+}
+
+void jaleco_vj_king_qtaro_device::map(address_map &map)
+{
+    map(0x10, 0x10).r(m_qtaro[0], FUNC(jaleco_vj_qtaro_device::reg2_r));
+    map(0x18, 0x1b).rw(m_qtaro[0], FUNC(jaleco_vj_qtaro_device::reg3_r), FUNC(jaleco_vj_qtaro_device::reg3_w));
+    map(0x20, 0x20).r(m_qtaro[1], FUNC(jaleco_vj_qtaro_device::reg2_r));
+    map(0x28, 0x2b).rw(m_qtaro[1], FUNC(jaleco_vj_qtaro_device::reg3_r), FUNC(jaleco_vj_qtaro_device::reg3_w));
+    map(0x30, 0x30).r(m_qtaro[2], FUNC(jaleco_vj_qtaro_device::reg2_r));
+    map(0x38, 0x3b).rw(m_qtaro[2], FUNC(jaleco_vj_qtaro_device::reg3_r), FUNC(jaleco_vj_qtaro_device::reg3_w));
+
+    map(0x50, 0x53).w(FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_requested_w<0>));
+    map(0x54, 0x57).w(FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_descriptor_phys_addr_w<0>));
+    map(0x58, 0x5b).rw(FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_running_r<0>), FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_running_w<0>));
+    map(0x60, 0x63).w(FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_requested_w<1>));
+    map(0x64, 0x67).w(FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_descriptor_phys_addr_w<1>));
+    map(0x68, 0x6b).rw(FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_running_r<1>), FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_running_w<1>));
+    map(0x70, 0x73).w(FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_requested_w<2>));
+    map(0x74, 0x77).w(FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_descriptor_phys_addr_w<2>));
+    map(0x78, 0x7b).rw(FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_running_r<2>), FUNC(jaleco_vj_king_qtaro_device::qtaro_dma_running_w<2>));
+
+    map(0x80, 0x83).rw(FUNC(jaleco_vj_king_qtaro_device::qtaro_fpga_firmware_status_r), FUNC(jaleco_vj_king_qtaro_device::qtaro_fpga_firmware_status_w));
+    map(0x84, 0x87).rw(FUNC(jaleco_vj_king_qtaro_device::qtaro_fpga_firmware_r), FUNC(jaleco_vj_king_qtaro_device::qtaro_fpga_firmware_w));
+    map(0x88, 0x8b).rw(FUNC(jaleco_vj_king_qtaro_device::king_qtaro_fpga_firmware_status_r), FUNC(jaleco_vj_king_qtaro_device::king_qtaro_fpga_firmware_status_w));
+    map(0x8c, 0x8f).rw(FUNC(jaleco_vj_king_qtaro_device::king_qtaro_fpga_firmware_r), FUNC(jaleco_vj_king_qtaro_device::king_qtaro_fpga_firmware_w));
```

Comment:
I'd reduce signature of these functions, it's implicit that they belongs to `jaleco_vj_king_qtaro_device` and it's also inconsistent. i.e. 
- `jaleco_vj_king_qtaro_device::king_qtaro_fpga_firmware_status_r` -> `jaleco_vj_king_qtaro_device::fpga_firmware_status_r`
- `jaleco_vj_king_qtaro_device::qtaro_dma_requested_w<2>` -> `jaleco_vj_king_qtaro_device::qtaro_dma_requested_w<2>`